Storage/SQL/BrowserDB.swift

```diff
@@ -64,229 +51,158 @@ open class BrowserDB {
-        let updateResult = schemaTable.update(conn, item: table, err: &err)
-        if  updateResult > 0 || insertResult > 0 {
-            return .updated
+        if let error = conn.setVersion(schema.version) {
+            log.error("Unable to update the schema version; \(error.localizedDescription)")
         }
-        return .failed
+
+        return true
     }
 
-    // Utility for table classes. They should call this when they're initialized to force
-    // creation of the table in the database.
-    func createOrUpdate(_ tables: Table...) -> DatabaseOpResult {
+    // Checks if the database schema needs created or updated and acts accordingly.
+    func prepareSchema(_ schema: Schema) -> DatabaseOpResult {
         guard !db.closed else {
-            log.info("Database is closed - skipping schema create/updates")
+            log.info("Database is closed; Skipping schema create or update.")
             return .closed
         }
 
-        var success = true
-
-        let doCreate = { (table: Table, connection: SQLiteDBConnection) -> Void in
-            switch self.createTable(connection, table: table) {
-            case .created:
-                success = true
-                return
-            case .exists:
-                log.debug("Table already exists.")
-                success = true
-                return
-            default:
-                success = false
-            }
+        // Get the current schema version for the database.
+        var currentVersion = 0
+        _ = self.db.withConnection(.readOnly, cb: { connection -> NSError? in
+            currentVersion = connection.version
+            return nil
+        })
+        
+        // If the current schema version for the database matches the specified
+        // `Schema` version, no further action is necessary and we can bail out.
+        if currentVersion == schema.version {
+            log.debug("Schema \(schema.name) already exists at version \(schema.version). Skipping additional schema preparation.")
+            return .success
         }
+        
+        log.debug("Schema \(schema.name) needs created or updated from version \(currentVersion) to \(schema.version).")
+
+        var success = true
 
         if let error = self.db.transaction({ connection -> Bool in
-            let thread = Thread.current.description
-            // If the table doesn't exist, we'll create it.
-            for table in tables {
-                log.debug("Create or update \(table.name) version \(table.version) on \(thread).")
-                if !table.exists(connection) {
-                    log.debug("Doesn't exist. Creating table \(table.name).")
-                    doCreate(table, connection)
+            // If the schema doesn't exist, we'll create it.
+            log.debug("Create or update \(schema.name) version \(schema.version) on \(Thread.current.description).")
+            if !schema.exists(connection) {
```

Comment:
No, because `user_version` will always be `0` for older versions. We're simply using this as a cue as to whether or not this is a brand new empty database or not. If its not, we want to call `schema.update()` instead of `schema.create()`.